cpp/src/c_api/renumber_arbitrary_edgelist.cu

```diff
@@ -0,0 +1,190 @@
+
+#include <cugraph/graph.hpp>
+#include <cugraph/utilities/error.hpp>
+
+#include <thrust/binary_search.h>
+#include <thrust/iterator/counting_iterator.h>
+
+namespace {
+
+template <typename vertex_t>
+cugraph_error_code_t renumber_arbitrary_edgelist(
+  raft::handle_t const& handle,
+  cugraph::c_api::cugraph_type_erased_host_array_view_t const* renumber_map,
+  cugraph::c_api::cugraph_type_erased_device_array_view_t* srcs,
+  cugraph::c_api::cugraph_type_erased_device_array_view_t* dsts)
+{
+  rmm::device_uvector<vertex_t> vertices(2 * srcs->size_, handle.get_stream());
+
+  thrust::copy_n(
+    handle.get_thrust_policy(), srcs->as_type<vertex_t>(), srcs->size_, vertices.data());
+  thrust::copy_n(handle.get_thrust_policy(),
+                 dsts->as_type<vertex_t>(),
+                 dsts->size_,
+                 vertices.data() + srcs->size_);
+
+  thrust::sort(handle.get_thrust_policy(), vertices.begin(), vertices.end());
+  vertices.resize(
+    thrust::distance(vertices.begin(),
+                     thrust::unique(handle.get_thrust_policy(), vertices.begin(), vertices.end())),
+    handle.get_stream());
+
+  vertices.shrink_to_fit(handle.get_stream());
+  rmm::device_uvector<vertex_t> ids(vertices.size(), handle.get_stream());
+  thrust::fill(handle.get_thrust_policy(),
+               ids.begin(),
+               ids.end(),
+               cugraph::invalid_vertex_id<vertex_t>::value);
+
+  raft::device_span<vertex_t const> vertices_span{vertices.data(), vertices.size()};
+  raft::device_span<vertex_t> ids_span{ids.data(), ids.size()};
+
+  // Read chunk of renumber_map in a loop, updating base offset to compute vertex id
+  // FIXME: Compute this as a function of free memory?  Or some value that keeps a
+  // particular GPU saturated?
+  size_t chunk_size = size_t{1} << 20;
+
+  rmm::device_uvector<vertex_t> renumber_chunk(chunk_size, handle.get_stream());
+
+  for (size_t chunk_base_offset = 0; chunk_base_offset < renumber_map->size_;
+       chunk_base_offset += chunk_size) {
+    size_t size = std::min(chunk_size, renumber_map->size_ - chunk_base_offset);
+    if (size < chunk_size) renumber_chunk.resize(size, handle.get_stream());
+
+    raft::update_device(renumber_chunk.data(),
+                        renumber_map->as_type<vertex_t>() + chunk_base_offset,
+                        size,
+                        handle.get_stream());
+
+    thrust::for_each(
+      handle.get_thrust_policy(),
+      thrust::make_counting_iterator<size_t>(0),
+      thrust::make_counting_iterator<size_t>(renumber_chunk.size()),
+      [chunk_base_offset,
+       renumber_chunk_span =
+         raft::device_span<vertex_t const>{renumber_chunk.data(), renumber_chunk.size()},
+       vertices_span,
+       ids_span] __device__(size_t idx) {
+        auto pos = thrust::lower_bound(
+          thrust::seq, vertices_span.begin(), vertices_span.end(), renumber_chunk_span[idx]);
+        if (pos != vertices_span.end()) {
```

Comment:
Shouldn't we check `&& *pos = renumber_chunk_span[idx]` as well?

Reply:
Yes.  Will fix.

Reply:
Just pushed update